src/main/java/io/github/phantamanta44/libnine/client/gui/component/impl/GuiComponentTextInput.java

```diff
@@ -149,16 +186,47 @@ public boolean onKeyPress(int keyCode, char typed) {
                 } else {
                     value = value.substring(0, currentLength - 1);
                 }
-                updateValidity();
-            } else if (keyCode == Keyboard.KEY_RETURN || keyCode == Keyboard.KEY_NUMPADENTER) {
-                updateValidity();
-                if (valid) callback.accept(value);
-            } else if (keyCode == Keyboard.KEY_ESCAPE) {
-                focused = false;
             }
+            updateValidityAndAccept();
             return true;
         }
         return false;
     }
 
+    private void onTextCharInput(char typed) {
+        switch (fieldType) {
+            case DIGITS_ONLY:
+                String val = value;
+                if (typed >= 48 && typed <= 57) { // digits 0..9
+                    if (typed == 48 && val.equals("")) return; // can't add "0" to empty
+                    val += typed;
+                } else if (typed == 107) { // letter "k", add 000 (multiply by thousand) if can
+                    if (val.equals("") || val.equals("0")) {
+                        val = "1000";
```

Comment:
this ignores length checks

Reply:
no big deal, it should be invalidated in that case